blog/2024-02-02-migrating-machines-to-xstate-v5/index.mdx

```diff
@@ -0,0 +1,651 @@
+
+:::warning
+
+Calling `setup().createMachine()` from XState v5 but then calling `assign` from v4 within a v5 machine could result in type errors.
+
+:::
+
+## 2. Move types to `setup()` and remove typegen
+
+One of the first migration steps, based on a breaking change, is to remove any TypeScript types from the `schema` property of the machine config from v4. These types should now be [included under a new `types` property](/docs/typescript#specifying-types) in the object passed to the new `setup({})` function.
+
+<Tabs>
+<TabItem value="v5" label="XState v5">
+
+```ts
+import { setup } from 'xstate';
+
+const machine = setup({
+  types: {} as {
+    context: {
+      /* type context props */
+    };
+    events: {
+      /* type events */
+    };
+  },
+  /* implementations */
+  actions: {},
+  guards: {},
+  actors: {},
+}).createMachine({
+  /* machine config */
+});
+```
+
+</TabItem>
+
+<TabItem value="v4" label="XState v4">
+
+```ts
+// ❌ DEPRECATED
+import { createMachine } from 'xstate';
+
+const machine = createMachine({
+  ...
+  tsTypes: {} as import('./myMachine.typegen').Typegen0,
+  schema: {
+    context: {} as {
+      /* type context props */
+    },
+    events: {} as {
+      /* type events */
+    },
+    services: {} as {
+      /* type services */
+    },
+  }
+});
+```
+
+</TabItem>
+</Tabs>
+
+When defining types, you may have noticed in past examples that an empty object is often cast as the types you want for each property like `events` and `actions`.
+
+```ts
+const machine = setup({
+  types: {
+    context: {} as {
+    };
+    events: {} as {
+    };
+  },
+})
+```
+
+That still works but but it’s even easier to just cast the whole types object all at once.
+
+```ts
+const machine = setup({
+  types: {} as {
+    context: {};
+    events: {};
+  },
+});
+```
+
+:::warning
+
+When chaining `setup().createMachine()` in v5, `types` can only be defined on the object passed to `setup()` and not on the config object passed to `.createMachine()`. Example: `setup({ types: {} }).createMachine({})`. A chained call to `createMachine()` only accepts the one argument, the machine config.
+
+:::
+
+### Remove typegen
+
+If you were previously using typegen in v4, you can also delete `tsTypes` as there is no more typegen in v5. Typing events in actions and guards can be done using type narrowing in those implementations. However, the following sections will show you how you can type a new `params` argument passed to those implementations.
+
+## 3. Convert action and guard strings to parameterized objects
+
+We must now [convert any named action or guard strings to action objects](/docs/migration#use-params-to-pass-params-to-actions--guards) which allows us to define an explicit `params` object (or primitive value) that will be received at runtime by our action implementations. There are two ways to do this:
```

Comment:
That is only a "requirement" when you depend on `event` because it helps you to adapt `event` into `params` within transition actions. If you have simple actions - or ones that only depend on `context` then you can continue using `entry: 'doStuff'` etc. But with `setup` you still have to declare `doStuff` in it.

Reply:
I'll definitely add some of the more nuanced considerations here and will make this a non "requirement".